File: Classes/Expressions/Parse/AJRExpressionParser.m

#import "AJRExpressionParser.h"

#import "AJRConstant.h"
#import "AJRExpression.h"
#import "AJRExpressionFunctionStackFrame.h"
#import "AJRExpressionStackFrame.h"
#import "AJRExpressionToken.h"
#import "AJRFormat.h"
#import "AJRFunction.h"
#import "AJROperator.h"
#import "AJRPlugInManager.h"
#import "NSCharacterSet+Extensions.h"
#import "NSString+Extensions.h"

static NSCharacterSet *whitespaceSet;
static NSMutableCharacterSet *literalStartSet;
static NSMutableCharacterSet *literalSet;
static NSCharacterSet *numberStartSet;
static NSCharacterSet *argumentNumberStartSet;
static NSCharacterSet *numberSet;
static NSMutableCharacterSet *operatorStartSet;
static NSMutableCharacterSet *operatorSet;

@interface AJRExpressionParser ()

- (AJRExpression *)expressionWithArguments:(va_list)arguments;

@end

@implementation AJRExpressionParser

#pragma mark Initialize

+ (void)initialize
{  
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        whitespaceSet = [NSCharacterSet whitespaceAndNewlineCharacterSet];
        argumentNumberStartSet = [NSCharacterSet characterSetWithCharactersInString:@"+-0123456789"];
        numberStartSet = [NSCharacterSet characterSetWithCharactersInString:@"0123456789"];
        numberSet = [NSCharacterSet characterSetWithCharactersInString:@"0123456789."];
        literalStartSet = [[NSCharacterSet ajr_swiftIdentifierStartCharacterSet] mutableCopy];
        literalSet = [[NSCharacterSet ajr_swiftIdentifierCharacterSet] mutableCopy];
        operatorStartSet = [[NSMutableCharacterSet alloc] init];
        operatorSet = [[NSMutableCharacterSet alloc] init];
        
        // Since we depend on the plug-in manager, make sure it's initialized.
        [AJRPlugInManager initializePlugInManager];
    });
}

#pragma mark Creation

- (id)initWithString:(NSString *)string
{
    return [self initWithStringFormat:string arguments:NULL];
}

- (id)initWithStringFormat:(NSString *)string, ...
{
    va_list     ap;
    
    va_start(ap, string);
    self = [self initWithStringFormat:string arguments:ap];
    va_end(ap);
    
    return self;
}

- (id)initWithStringFormat:(NSString *)string arguments:(va_list)arguments
{
    if ((self = [super init])) {
        if ((string == nil) || ([[string stringByTrimmingCharactersInSet:whitespaceSet] length] == 0)) {
            @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"Unable to create AJRExpression. No expression string supplied." userInfo:nil];
        }
        
        _input = [string dataUsingEncoding:NSUTF32StringEncoding];
        _rawInput = [_input bytes];
        _length = [_input length] / sizeof(UTF32Char);
        _position = 1; // Skip the BOM marker.
        // Fault the expression immediately. This is necessary, because we have the arguments available.
        [self expressionWithArguments:arguments];
    }
    return self;
}

#pragma mark - Utilities

- (NSString *)stringFromRange:(NSRange)range
{
    UTF32Char buffer[range.length + 1];
    buffer[0] = _rawInput[0]; // Make sure to use the same BOM.
    memcpy(buffer + 1, _rawInput + range.location, range.length * sizeof(UTF32Char));
    NSData *data = [[NSData alloc] initWithBytesNoCopy:buffer length:(range.length + 1) * sizeof(UTF32Char) freeWhenDone:NO];
    return [[NSString alloc] initWithData:data encoding:NSUTF32StringEncoding];
}

#pragma mark Parse

- (void)readWhitespace
{
    while ((_position < _length) && [whitespaceSet longCharacterIsMember:_rawInput[_position]]) {
        _position++;
    }
}

- (AJRExpressionToken *)readNumber
{
    NSInteger    start = _position;
    BOOL        hasDecimal = NO;
    NSNumber    *value;
    
    // Make sure we move over a +/-
    while (_position < _length) {
        UTF32Char character = _rawInput[_position];
        
        if (![numberSet longCharacterIsMember:character]) break;
        if (character == '.') {
            if (hasDecimal) break;
            hasDecimal = YES;
        }
        
        _position++;
    }
    
    value = (id)[self stringFromRange:(NSRange){start, _position - start}];
    if (hasDecimal) {
        value = [NSNumber numberWithDouble:[value floatValue]];
    } else {
        value = [NSNumber numberWithLong:[value longValue]];
    }
    
    return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeNumber value:value];
}

- (AJRExpressionToken *)readOperator
{
    NSInteger start = _position;
    NSString *value;
    AJROperator *operator;
    
    while (_position < _length && [operatorSet longCharacterIsMember:_rawInput[_position]]) {
        _position++;
    }
    
    // Get the token from the stream
    value = [self stringFromRange:(NSRange){start, _position - start}];
    
    // See if it's an operator
    operator = [AJROperator operatorForToken:value];
    if (operator) {
        return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeOperator value:operator];
    }
    
    // Fell through, so treat it as a literal. Note, we'll never be a constant, like we have in the readLiteral code, because constants are registered as literals, which means having a constant will cause us to enter the readLiteral code rather than the readOperator code.
    return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeLiteral value:value];
}

- (AJRExpressionToken *)readLiteral
{
    NSInteger start = _position;
    NSString *value;
    AJROperator *operator;
    AJRConstant *constant;
    
    while (_position < _length && ([literalSet longCharacterIsMember:_rawInput[_position]] || _rawInput[_position] == '.' || _rawInput[_position] == '@')) {
        _position++;
    }
    
    // Get the token from the stream
    value = [self stringFromRange:(NSRange){start, _position - start}];
    
    if (_position < _length && _rawInput[_position] == '(') {
        // We have a function declaration.
        Class functionClass = [AJRFunction functionClassForName:value];
        
        // Consume the opening parenthesis.
        _position++;
        
        if (functionClass) {
            AJRFunction *function = [[functionClass alloc] init];
            AJRExpressionToken *token = [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeFunction value:function];
            return token;
        }
        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:AJRFormat(@"Unknown function: %@", value) userInfo:nil];
    }
                
    // See if it's an operator
    operator = [AJROperator operatorForToken:value];
    if (operator) {
        return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeOperator value:operator];
    }
    constant = [AJRConstant constantForToken:value];
    if (constant) {
        return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeNumber value:constant];
    }
    
    return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeLiteral value:value];
}

#define INCREMENT_POSITION() { \
    outPosition++; \
    if (outPosition == max) { \
        max += 32; \
        buffer = (UTF32Char *)NSZoneRealloc(NULL, buffer, sizeof(UTF32Char) * max); \
    } \
}

- (AJRExpressionToken *)readStringWithStartCharacter:(UTF32Char)startCharacter
{  
    AJRExpressionToken *token;
    UTF32Char *buffer;
    NSInteger max = 32;
    NSInteger outPosition = 0;
    UTF32Char character;
    
    buffer = (UTF32Char *)NSZoneMalloc(nil, sizeof(UTF32Char) * (max + 1));
    buffer[outPosition++] = _rawInput[0];
    
    _position++; // skip past the opening quote
    while (_position < _length) {
        character = _rawInput[_position];
        if (character == startCharacter) {
            break; 
        } else if (character == '\\') {
            _position++;
            if (_position >= _length) break;
            character = _rawInput[_position];
            if (character == 'n') {
                character = '\n'; 
            } else if (character == 'r') {
                character = '\r';
            } else if (character == 'e') {
                character = '\e';
            } else if (character == 't') {
                character = '\t';
            } else if (character == 's') {
                character = ' ';
            } else if (character == '\'') {
                character = '\'';
            } else if (character == '"') {
                character = '"';
            } else if (character == '\\') {
                character = '\\';
            } else {
                character = '?';
            }
            buffer[outPosition] = character;
            INCREMENT_POSITION();
        } else {
            buffer[outPosition] = character;
            INCREMENT_POSITION();
        }
        _position++;
    }
    
    NSData *data = [[NSData alloc] initWithBytesNoCopy:buffer length:outPosition * sizeof(UTF32Char) freeWhenDone:NO];
    NSString *string = [[NSString alloc] initWithData:data encoding:NSUTF32StringEncoding];
    token = [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeString value:string];
    NSZoneFree(nil, buffer);
    _position++;
    
    return token;
}

- (AJRExpressionToken *)tokenForValue:(id)value
{
    AJRExpressionToken *token = nil;
    
    if (value == nil) {
        token = [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeNumber value:nil];
    } else if ([value isKindOfClass:[NSNumber class]]) {
        token = [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeNumber value:value];
    } else {
        // A string, or something we're going to treat as a string.
        NSString *stringValue = [value description];
        AJRConstant *constant = [AJRConstant constantForToken:stringValue];
        if (constant) {
            token = [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeNumber value:constant];
        } else {
            token = [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeString value:stringValue];
        }
    }
    
    return token;
}

- (AJRExpressionToken *)readArgumentExpandingConstants:(BOOL)expandConstants withArguments:(va_list)arguments
{
    UTF32Char character;
    
    if (_position >= _length) {
        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"No modifier to %%" userInfo:nil];
    }
    
    character = _rawInput[_position];
    _position++;
    if (character == 'd') {
        NSInteger arg = va_arg(arguments, NSInteger);
        return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeNumber value:[NSNumber numberWithInteger:arg]];
    } else if (character == 's') {
        const char *value = va_arg(arguments, char *);
        return [self tokenForValue:value ? [NSString stringWithCString:value encoding:NSUTF8StringEncoding] : nil];
    } else if (character == '@') {
        return [self tokenForValue:va_arg(arguments, id)];
    } else if (character == 'f') {
        double arg = va_arg(arguments, double);
        return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeNumber value:[NSNumber numberWithDouble:arg]];
    } else {
        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:AJRFormat(@"No modifier to %%: %c", character) userInfo:nil];
    }
    
    return nil;
}

- (AJRExpressionToken *)nextTokenWithArguments:(va_list)arguments
{
    UTF32Char character;
    
    // Ignore any leading whitespace
    [self readWhitespace];
    
    if (_position < _length) {
        character = _rawInput[_position];
        if (character == '(') {
            _position++;
            return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeOpenParen];
        } else if (character == ')') {
            _position++;
            return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeCloseParen];
        } else if (character == ',') {
            _position++;
            return [AJRExpressionToken tokenWithType:AJRExpressionTokenTypeComma];
        } else if ([numberStartSet longCharacterIsMember:character]) {
            return [self readNumber];
        } else if (character == '"' || character == '\'') {
            return [self readStringWithStartCharacter:character];
        } else if (character == '%') {
            _position++;
            return [self readArgumentExpandingConstants:YES withArguments:arguments];
        } else if ([literalStartSet longCharacterIsMember:character]) {
            // Anything not identified above is a literal, which might turn out to be an operator
            // or a key, or something else.
            return [self readLiteral];
        } else if ([operatorStartSet longCharacterIsMember:character]) {
            return [self readOperator];
        } else {
            @throw [NSException exceptionWithName:NSInvalidArgumentException reason:AJRFormat(@"Unexpected character in input: 0x%x '%c'", character, character) userInfo:nil];
        }
    }
    
    return nil;
}

- (AJRExpression *)expressionWithArguments:(va_list)arguments
{
    if (_expression == nil) {
        AJRExpressionToken *token;
        AJRExpressionStackFrame *frame;
        
        _stack = [[NSMutableArray alloc] init];
        [_stack addObject:[AJRExpressionStackFrame frame]];
        
        while ((token = [self nextTokenWithArguments:arguments])) {
            // Used by a lot below...
            frame = [_stack lastObject];
            
            switch ([token type]) {
                case AJRExpressionTokenTypeString:
                case AJRExpressionTokenTypeNumber:
                case AJRExpressionTokenTypeLiteral:
                case AJRExpressionTokenTypeOperator:
                    if ([[token value] isKindOfClass:[AJRExpression class]]) {
                        [frame addExpression:[token value]];
                    } else {
                        [frame addToken:token];
                    }
                    break;
                case AJRExpressionTokenTypeOpenParen:
                    [_stack addObject:[AJRExpressionStackFrame frame]];
                    break;
                case AJRExpressionTokenTypeCloseParen:
                    if ([_stack count] <= 1) {
                        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"Unbalanced parentheses in expression" userInfo:nil];
                    } else {
                        AJRExpression    *expression;
                        // Make sure this doesn't free itself when we remove it from the _stack.
                        [_stack removeLastObject];
                        // And add the subframe's expression to the preceeding stack frame.
                        expression = [frame expression];
                        [expression setProtected:YES];
                        [[_stack lastObject] addExpression:expression];
                    }
                    break;
                case AJRExpressionTokenTypeFunction:
                    frame = [AJRExpressionFunctionStackFrame frameWithFunction:(AJRFunction *)[token value]];
                    [_stack    addObject:frame];
                    break;
                case AJRExpressionTokenTypeComma: {
                    // Search the stack frame for a function expression
                    for (NSInteger x = [_stack count] - 1; x >= 0; x--) {
                        AJRExpressionFunctionStackFrame    *frame = [_stack objectAtIndex:x];
                        if ([frame isKindOfClass:[AJRExpressionFunctionStackFrame class]]) {
                            [frame reduceArgument];
                        }
                    }
                    break;
                }
            }
        }
        
        // Modified the below line to check for one than one item on the _stack.
        // If no parenthesis
        // were used then there will only be one item on the _stack.
        if ([_stack count] > 1) {
            // frame = _stack.get(_stack.size() - 1);
            // _stack.remove(_stack.size() - 1);
            // _stack.get(_stack.size() - 1).applyFrame(frame);
            // This should be an error condition, because it means we opened a
            // parenthesis, but
            // didn't close it.
            @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"Illegal expression string, probably caused by an unclosed parenthesis." userInfo:nil];
        }
        
        _expression = [[_stack lastObject] expression];
        _stack = nil;
    }
    
    return _expression;
}

- (AJRExpression *)expression
{
    return _expression;
}

+ (AJRExpression *)expressionForString:(NSString *)string
{
    return [[[AJRExpressionParser alloc] initWithString:string] expression];
}

+ (AJRExpression *)expressionForStringFormat:(NSString *)string, ...
{
    AJRExpression *expression;
    va_list ap;
    
    va_start(ap, string);
    expression = [self expressionForStringFormat:string arguments:ap];
    va_end(ap);
    
    return expression;
}

+ (AJRExpression *)expressionForStringFormat:(NSString *)string arguments:(va_list)arguments
{
    return [[[AJRExpressionParser alloc] initWithStringFormat:string arguments:arguments] expression];
}

#pragma mark Tokens

+ (void)addLiteralToken:(NSString *)token
{
    if ([token length]) {
        NSData *data = [token dataUsingEncoding:NSUTF32StringEncoding];
        const UTF32Char *buffer = [data bytes];
        NSInteger length = [data length] / sizeof(UTF32Char);
        for (NSInteger x = 1; x < length; x++) {
            if (x == 1) {
                [literalStartSet addCharactersInRange:(NSRange){buffer[x], 1}];
            }
            [literalSet addCharactersInRange:(NSRange){buffer[x], 1}];
        }
    }
}

+ (void)addOperatorToken:(NSString *)token
{
    if ([token rangeOfCharacterFromSet:literalStartSet].location != NSNotFound) {
        [self addLiteralToken:token];
    } else if ([token length]) {
        NSData *data = [token dataUsingEncoding:NSUTF32StringEncoding];
        const UTF32Char *buffer = [data bytes];
        NSInteger length = [data length] / sizeof(UTF32Char);
        for (NSInteger x = 1; x < length; x++) {
            if (x == 1) {
                [operatorStartSet addCharactersInRange:(NSRange){buffer[x], 1}];
            }
            [operatorSet addCharactersInRange:(NSRange){buffer[x], 1}];
        }
    }
}

@end

File: Classes/Expressions/Parse/AJRExpressionParser.swift
/*
 AJRExpressionParser.swift
 AJRFoundation

 Copyright © 2021, AJ Raftis and AJRFoundation authors
 All rights reserved.

 Redistribution and use in source and binary forms, with or without modification,
 are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.
 * Neither the name of AJRFoundation nor the names of its contributors may be
   used to endorse or promote products derived from this software without
   specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL AJ RAFTIS BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import Foundation

public enum AJRExpressionParserError : Error {
    
    case invalidToken(String)
    case unexpectedTokenSequence(String)
    case invalidReductionState(String)
    case failedToFullyReduce(String)
    case unknownFunction(String)
    case invalidState(String)
    case missingModifier(String)
    case invalidModifier(String)
    case insufficientArguments(String)
    case invalidType(String)
    case invalidCharacter(String)
    case unbalancedParentheses(String)
    case invalidInput(String)

}

@objcMembers
public class AJRExpressionParser : NSObject {

    // MARK: - Character Sets
    
    static var whitespaceSet = CharacterSet.whitespacesAndNewlines
    static var argumentNumberStartSet = CharacterSet(charactersIn: "+-0123456789")
    static var numberStartSet = CharacterSet(charactersIn: "0123456789")
    static var numberSet = CharacterSet(charactersIn: "0123456789.")
    static var literalStartSet = CharacterSet.swiftIdentifierStartCharacterSet
    static var literalSet = CharacterSet.swiftIdentifierCharacterSet
    static var operatorStartSet = CharacterSet()
    static var operatorSet = CharacterSet()

    // MARK: - Properties
    
    public var string : String
    public var length : String.Index
    public var position : String.Index
    public var stack : [AJRExpressionStackFrame]
    public var arguments : [Any?]
    public var argumentIndex : Int
    internal var _expression: AJRExpression?
    
    // MARK: - Creation
    
    public init(string: String) throws {
        AJRExpression.initialize() // Make sure all the operators, functions, and constants are registered
        self.string = string
        self.length = self.string.endIndex
        self.position = self.string.startIndex
        self.stack = [AJRExpressionStackFrame]()
        self.arguments = []
        self.argumentIndex = 0
    }

    @objc (initWithFormat:arguments:error:)
    public convenience init(format: String, _ arguments: [Any]) throws {
        try self.init(string: format)
        self.arguments = arguments
    }

    public convenience init(format: String, _ arguments: [Any?]) throws {
        try self.init(string: format)
        self.arguments = arguments
    }
    
    // MARK: - Utilities
    
    internal func nextArgument<T>() throws -> T? {
        if argumentIndex >= arguments.count {
            throw AJRExpressionParserError.insufficientArguments("Not enough arguments were provided by caller.")
        }
        let argument = arguments[argumentIndex];
        argumentIndex += 1
        if argument == nil || (argument as? NSObject) == NSNull.init() {
            return nil
        }
        if let argument = argument as? T {
            return argument
        }
        throw AJRExpressionParserError.invalidType("Argument was not of a valid type: \(type(of:T.self))")
    }
    
    // MARK: - Parse
    
    internal func readWhitespace() throws -> Void {
        while (position < length) && AJRExpressionParser.whitespaceSet.contains(string[position]) {
            position = string.index(after: position)
        }
    }
    
    internal enum ForcedType : Character, CaseIterable {
        case none = "\u{0000}"
        case integer = "i"
        case float = "f"
        case hour = "H"
        case minute = "M"
        case second = "S"
        case month = "m"
        case day = "d"
        case year = "y"
    }
    
    internal func readNumber() throws -> AJRExpressionToken {
        let start = position
        var hasDecimal = false
        
        // Make sure we move over a +/-
        while position < length {
            let character = string[position]
            
            if !AJRExpressionParser.numberSet.contains(character) {
                break
            }
            if character == "." {
                // Make sure to only read on decimal point
                if hasDecimal {
                    break
                }
                hasDecimal = true
            }
            
            position = string.index(after: position)
        }
        
        // See if we have a modifier
        var forcedType = ForcedType.none
        if position < length {
            let character = string[position]
            for type in ForcedType.allCases {
                if character == type.rawValue {
                    forcedType = type
                    break
                }
            }
        }

        // Grab the numeric portion of the string
        let substring = string[start..<position]

        // See if we need to consume the "type" charcter
        if forcedType != .none {
            position = string.index(after: position)
        }

        let value : Any
        let numberType : AJRExpressionToken.TokenType
        // And create our value
        switch forcedType {
        case .none:
            if hasDecimal {
                value = Double(substring)!
            } else {
                value = Int(substring)!
            }
            numberType = .number
        case .integer:
            value = Int(substring)!
            numberType = .number
        case .float:
            value = Int(substring)!
            numberType = .number
        case .hour:
            value = DateComponents(hour: Int(substring))
            numberType = .dateComponent
        case .minute:
            value = DateComponents(minute: Int(substring))
            numberType = .dateComponent
        case .second:
            value = DateComponents(second: Int(substring))
            numberType = .dateComponent
        case .day:
            value = DateComponents(day: Int(substring))
            numberType = .dateComponent
        case .month:
            value = DateComponents(month: Int(substring))
            numberType = .dateComponent
        case .year:
            value = DateComponents(year: Int(substring))
            numberType = .dateComponent
        }

        return AJRExpressionToken.token(type: numberType, value: value)
    }

    internal func readOperator() throws -> AJRExpressionToken {
        let start = position
        
        while position < length && AJRExpressionParser.operatorSet.contains(string[position]) {
            position = string.index(after:position)
        }
        
        // Get the token from the stream
        let stringValue = String(string[start..<position])
        
        // See if it's an operator
        if let `operator` = AJROperator.operatorForToken(stringValue) {
            return AJRExpressionToken.token(type: .operator, value:`operator`)
        }
        
        // Fell through, so treat it as a literal. Note, we'll never be a constant, like we have in the readLiteral code, because constants are registered as literals, which means having a constant will cause us to enter the readLiteral code rather than the readOperator code.
        return AJRExpressionToken.token(type: .literal, value: stringValue)
    }
    
    internal func readLiteral() throws -> AJRExpressionToken {
        let start = position
    
        while position < length && (AJRExpressionParser.literalSet.contains(string[position]) || string[position] == "." || string[position] == "@") {
            position = string.index(after: position)
        }
    
        // Get the token from the stream
        let stringValue = String(string[start..<position])
    
        if position < length && string[position] == "(" {
            // We have a function declaration.
            let function = AJRFunction.function(for: stringValue)
    
            // Consume the opening parenthesis.
            position = string.index(after: position)
    
            if let function = function {
                return AJRExpressionToken.token(type: .function, value: function.copy())
            }
            throw AJRExpressionParserError.unknownFunction("Unknown function: \(stringValue)")
        }
    
        // See if it's an operator
        if let `operator` = AJROperator.operatorForToken(stringValue) {
            return AJRExpressionToken.token(type: .operator, value: `operator`)
        }
        if let constant = AJRConstant.constant(forToken: stringValue) {
            return AJRExpressionToken.token(type: .number, value: constant)
        }
    
        return AJRExpressionToken.token(type: .literal, value: stringValue)
    }

    public func readString(startCharacter: Character) throws -> AJRExpressionToken {
        var buffer = ""
        
        position = string.index(after: position) // skip past the opening quote
        while position < length {
            var character = string[position]
            if character == startCharacter {
                // Skip past the closing character
                position = string.index(after: position)
                break
            } else if character == "\\" {
                position = string.index(after: position)
                if position >= length {
                    break
                }
                character = string[position]
                if character == "n" {
                    character = "\n"
                } else if character == "r" {
                    character = "\r"
                } else if character == "e" {
                    character = "\u{1b}"
                } else if character == "t" {
                    character = "\t"
                } else if character == "s" {
                    character = " "
                } else if character == "'" {
                    character = "'"
                } else if character == "\"" {
                    character = "\""
                } else if character == "\\" {
                    character = "\\"
                } else {
                    character = "?"
                }
            }
            buffer.append(character)
            position = string.index(after: position)
        }
    
        return AJRExpressionToken.token(type: .string, value: buffer)
    }

    public func token(forValue value: Any?) throws -> AJRExpressionToken {
        var token: AJRExpressionToken? = nil
        
        if value == nil {
            token = AJRExpressionToken.token(type: .number, value: nil)
        } else if value is (any BinaryInteger) || value is (any BinaryFloatingPoint) {
            token = AJRExpressionToken.token(type: .number, value: value)
        } else if value is NSNumber {
            // This means we got a value via the Obj-C bridge.
            token = AJRExpressionToken.token(type: .number, value: value)
        } else if let value = value as? CustomStringConvertible {
            // A string, or something we're going to treat as a string.
            let stringValue = value.description
            if let constant = AJRConstant.constant(forToken: stringValue) {
                token = AJRExpressionToken.token(type: .number, value: constant)
            } else {
                token = AJRExpressionToken.token(type: .string, value: stringValue)
            }
        } else {
            throw AJRExpressionParserError.invalidState("Found a value we couldn't handle. This shouldn't happen: \(value!)")
        }
        
        return token!
    }
    
    internal func readArgument(expandingConstants expandConstants: Bool) throws -> AJRExpressionToken? {
        if position >= length {
            throw AJRExpressionParserError.missingModifier("No modifier to %%")
        }
    
        let character = string[position]
        position = string.index(after: position)
        if character == "d" {
            let arg : Int? = try nextArgument()
            return AJRExpressionToken.token(type: .number, value: arg)
        } else if character == "s" {
            let arg : String? = try nextArgument()
            return try token(forValue: arg)
        } else if character == "@" {
            let arg : AnyObject? = try nextArgument()
            return try token(forValue: arg)
        } else if character == "f" {
            let arg : Double? = try nextArgument()
            return AJRExpressionToken.token(type: .number, value: arg)
        } else {
            throw AJRExpressionParserError.invalidModifier("No modifier to %%: \(character)")
        }
    }

    public func nextToken() throws -> AJRExpressionToken? {
        // Ignore any leading whitespace
        try readWhitespace()
        
        if position < length {
            let character = string[position]
            if character == "(" {
                position = string.index(after: position)
                return AJRExpressionToken.token(type: .openParen)
            } else if character == ")" {
                position = string.index(after: position)
                return AJRExpressionToken.token(type: .closeParen)
            } else if character == "," {
                position = string.index(after: position)
                return AJRExpressionToken.token(type: .comma)
            } else if AJRExpressionParser.numberStartSet.contains(character) {
                return try readNumber()
            } else if character == "\"" || character == "'" {
                return try readString(startCharacter: character)
            } else if character == "%" {
                position = string.index(after: position)
                return try readArgument(expandingConstants: true)
            } else if AJRExpressionParser.literalStartSet.contains(character) {
                // Anything not identified above is a literal, which might turn out to be an operator
                // or a key, or something else.
                return try readLiteral()
            } else if AJRExpressionParser.operatorStartSet.contains(character) {
                return try readOperator()
            } else {
                throw AJRExpressionParserError.invalidCharacter("Unexpected character in input: \(character.unicodeScalars) '\(character)'")
            }
        }
        
        return nil
    }

    @objc(expressionWithError:)
    public func expression() throws -> AJRExpression {
        if _expression == nil {
            stack = [AJRExpressionStackFrame]()
            stack.append(AJRExpressionStackFrame())
    
            while let token = try nextToken() {
                // Used by a lot below...
                let frame = stack.last!
    
                switch token.type {
                    
                case .string: fallthrough
                case .number: fallthrough
                case .literal: fallthrough
                case .dateComponent: fallthrough
                case .operator:
                    if let value = token.value as? AJRExpression {
                        try frame.add(expression: value)
                    } else {
                        try frame.add(token: token)
                    }

                case .openParen:
                    stack.append(AJRExpressionStackFrame())
                    
                case .closeParen:
                    if stack.count <= 1 {
                        throw AJRExpressionParserError.unbalancedParentheses("Unbalanced parentheses in expression")
                    } else {
                        // Make sure this doesn't free itself when we remove it from the _stack.
                        stack.removeLast()
                        // And add the subframe's expression to the preceeding stack frame.
                        let expression : AJRExpression = try frame.expression()
                        expression.protected = true
                        try stack.last!.add(expression: expression)
                    }
                    
                case .function:
                    stack.append(AJRExpressionFunctionStackFrame(function: token.value as! AJRFunction))

                case .comma:
                    // Search the stack frame for a function expression
                    for index in stride(from: stack.count - 1, to: 0, by: -1) {
                        if let frame = stack[index] as? AJRExpressionFunctionStackFrame {
                            try frame.reduceArgument()
                            break;
                        }
                    }
                }
            }
    
            // Modified the below line to check for one than one item on the _stack.
            // If no parenthesis
            // were used then there will only be one item on the _stack.
            if stack.count > 1 {
                // frame = _stack.get(_stack.size() - 1);
                // _stack.remove(_stack.size() - 1);
                // _stack.get(_stack.size() - 1).applyFrame(frame);
                // This should be an error condition, because it means we opened a
                // parenthesis, but
                // didn't close it.
                throw AJRExpressionParserError.invalidInput("Illegal expression string, probably caused by an unclosed parenthesis.")
            }
    
            _expression = try stack.last!.expression()
            stack.removeAll()
        }
    
        // We either threw an error, or expression is now initialized
        return _expression!
    }
    
    // MARK: - Building Expressions

    @objc(expressionForString:error:)
    public class func expression(string: String) throws -> AJRExpression {
        return try AJRExpressionParser(string: string).expression()
    }

    @objc(expressionWithFormat:arguments:error:)
    public class func expression(format: String, _ arguments: [Any]) throws -> AJRExpression {
        return try AJRExpressionParser(format: format, arguments).expression()
    }

    public class func expression(format: String, _ arguments: Any?...) throws -> AJRExpression {
        return try AJRExpressionParser(format: format, arguments).expression()
    }

    // MARK: - Literals
    
    public class func addLiteralToken(_ token: String) -> Void {
        if token.count > 0 {
            literalSet.insert(charactersIn: token)
            literalStartSet.insert(character: token[token.startIndex])
        }
    }
    
    public class func addOperatorToken(_ token: String) -> Void {
        if token.rangeOfCharacter(from: literalStartSet) != nil {
            addLiteralToken(token)
        } else if token.count > 0 {
            operatorSet.insert(charactersIn: token)
            operatorStartSet.insert(character: token[token.startIndex])
        }
    }
    
}
